### ISSUE TITLE
Streaming response blocks worker from handling other requests while streaming to client

### ISSUE THREAD

**Issue by LinusCDE** [actix/actix-web#2774]
## Expected Behavior
Streaming responses will behave like normal responses in regards to connection limits.

## Current Behavior
Running streamed responses can make other requests fail, timeout or stall.

My guess is that the streamed response somehow prevents the current worker from handling anymore requests while the response is being streamed. This shouldn't happen since the response is streamed in an async fashion.

## Possible Solution
Not sure, but I would suspect some bug in how streamed responses are handled. I highly doubt that this would be intended behaviour.

I searched for the bug and didn't find any record of it. `max_connections()` and `max_connection_rate()` is already high and setting both explicitly to 100 didn't change the behaviour at all.

When enough streaming responses are stopped, new request that stalled usually also return a 408 (timed out), which seems to suggest that the request may have technically been started / timed internally to notice the timeout when handling it.

## Steps to Reproduce (for bugs)

1. Set a low amount of workers (1 or 2)
2. Have one or 2 streamed responses running
3. Other requests will start to take a while, timeout or stall (connected and new web sockets are also affected)

## Context
Using actix-web here to download a large usb image. This download can take long and I noticed that having a download running makes other webpages (partially) unresponsive. Sometimes only websockets stopped connecting, stalled or even simple requests didn't go through as long as the download was running.

I'm using 2 workers as more makes no sense on the target platform (weak singlecore arm cpu). Increasing the amount "fixes" the problem. Running more downloads at once eventually stall the rest again though, making this an easy way to dos the server.

## Your Environment

- Rust Version (I.e, output of `rustc -V`): rustc 1.63.0-nightly (12cd71f4d 2022-06-01)
- Actix Web Version: 4.0.1
- Platform: Was running on armv7, but also does the same on x86_64 (host)

I recently migrated from an old actix 3.X to 4.0.1. I didn't notice this behaviour before, so this might be introduced in the last 2 years (or 3.X -> 4; no guarantee on this).

Below is a video demonstrating the bug manifesting:

https://user-images.githubusercontent.com/22298664/171999815-c3793d3a-17dc-4910-a432-7da9727ef237.mp4

**Comment by LinusCDE** [comment]
I tried to tackle the problem again today and found at least a way to do it without the worker locking up.

It seems that using `BodyStream` instead of `SizedStream`. I found `BodyStream` when search for a possible alternative to `SizedStream` in the same module in the docs [here](https://docs.rs/actix-web/latest/actix_web/body/index.html). The description of `SizedStream` vs `BodyStream` tipped me off, as it seemed that `SizedStream` doesn't do chunking, which is the default for streaming responses I thought.

So it seems that `SizedStream` is supposed to be used if a response has a known size. It is doesn't in a stream like fashion, but also not quite. And somehow because of that, responding somehow puts Actix in a weird state, where it doesn't realize the worker as active and still issues requests to it which can then lock up until the stream for `SizedStream` is done.

Maybe the intention of `SizedStream` to have a stream that takes a bit longer and prevents filling up RAM, but also not be that long to process. I certainly misunderstood the intentions and the docus for `SizedStream` should probably be updated to warn users of this happening.

Also I now seem to have another issue: While my response takes a long time, I actually know the size. I can't add the header `Content-Length` myself to `BodyStream` as the internally returned `BodySize::Stream` causes it to be overridden / removed.

So I actually need kinda a hybrid between `BodyStream` and `SizedStream`. Where it doesn't queue requests on the same worker, as the response can take many minutes, but also have the ability to specify the size.

To sum it up:
 - Please specify in the docs that `SizedStream` causes new requests to be queued up in the same worker and potentially time out if the response takes a long time to stream/download to the client. This was a behaviour I didn't expect at all and it confused me for a very long time and was a potential breaking point with the web server for me, as downloading a large amount of data this way bricked the webserver (more pronounced the less workers where available or the more downloads where running at once, fully bricked when running downloads == worker count).
 - Add some way to have a BodyStream with a defined / expected size.

**Comment by LinusCDE** [comment]
Thought I found the issue (using `SizedStream` instead of `BodyStream`) but more testing revealed I just had luck with the distribution of requests accross the workers and the same problem persists with `BodyStream`.

**Comment by LinusCDE** [comment]
My current assumption is, that returning not enough `Poll:Pending`s freezes the actix worker somehow. I'm using [StreamReader](https://docs.rs/tokio-util/latest/tokio_util/io/struct.StreamReader.html) to turn a `AsyncRead` into a stream. I looked into the implementation of that, and it seems pretty sane.

My assumption know is that actix-files ChunkedFileRead somehow has a Poll-Pattern which Actix has no problem with, but other Async things just cause problem. I'm pretty much stumped on this problem. At probably need to somehow re-implement many things actix files is doing.

Tested with `.body(BodyStream::new(...))`, `.body(SizedStream::new(...))` and `.streaming(...)`. All seem to have the same issue.

**Comment by robjtede** [comment]
My immediate feeling was about the single core ARM chip having resource limitations but you seem to suggest you can show this on your development machine too. So, I want to look into this. Are you able to give me a reproduction repo that shows the issue on your dev machine?

**Comment by EnderKill98** [comment]
I can probably make some minimal example for it. And yes it works on both the arm device and my host pc. The behaviour only changes depending on how many workers are used. More workers seem to kinda mitigate it, since it seem the requests are more likely to be handle on workers that are not busy responding.

**Comment by LinusCDE** [comment]
Sure, I created a working example here: https://github.com/LinusCDE/actix-streaming-response-issue-demo

It seems to behave the same. For testing purposes it generates a sequence of bytes from 0 to 255 and repeats them until a certain size is reached (here 0,5 GiB). I used this for testing on my dev machiene. The real thing is some Linux file (not a regular one, but can be read like one). They both behave the same and I left the code in to potentially test the real file as well.

Just compile and run it, start up to 4 downloads of http://localhost:8080/download and try to load the index page while those are running.

This doesn't seem to depend on the machine. Sure the ARM machine is pretty weak, but the problem seems to depend on the workers set.

**Comment by LinusCDE** [comment]
Hi,

decided to re-test this today and it seems that an update to actix_web 4.2 seems to have mitigated or maybe even solved this issue.

When I now try to download as many files as workers, the webserver still becomes really slow to respond to frontend reloads (lots of js, css, etc. requests), but it doesn't seem to full hang anymore.

In addition I found some example code to make a simple `tokio_util::io::ReaderStream` myself on [SO here](https://stackoverflow.com/a/64486497). I heavily adapted it and it seems to lessen the hangs a fair bit since I can control the buffer size myself.

```rust
use pin_project_lite::pin_project;
use std::pin::Pin;
use std::task::{Context, Poll};
use tokio::io::{AsyncRead, ReadBuf};

pin_project! {
    pub struct AsyncReadStreamer<T: AsyncRead> {
        #[pin]
        pub inner: T,
    }
}

impl<T: AsyncRead> tokio_stream::Stream for AsyncReadStreamer<T> {
    type Item = Result<actix_web::web::Bytes, actix_web::Error>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.project();

        let mut raw_buffer = [0; 1024 * 256]; // 256 KiB
        let mut buffer = ReadBuf::new(&mut raw_buffer);

        match this.inner.poll_read(cx, &mut buffer) {
            Poll::Ready(Ok(_)) => Poll::Ready(Some(Ok(actix_web::web::Bytes::copy_from_slice(buffer.filled())))),
            Poll::Ready(Err(err)) => Poll::Ready(Some(Err(actix_web::error::ErrorInternalServerError(err)))),
            Poll::Pending => Poll::Pending,
        }
    }
}
```

Since I don't seem to be able to fully block workers anymore, I consider this fixed. The remaining lag is probably just the fact, that the device is a slow arm processor and also really I/O-constrained while streaming a response multiple times. I can simply limit the amount of concurrent downloads of the file to e.g. 50% of the worker threads to make requests hardly ever hang.

So I consider this solved. `tokio_util::io::ReaderStream` seems less than ideal imo, so whoever stumbles upon the same issue, can use the code above to improve performance ig.

EDIT: Take my replacement for `tokio_util::io::ReaderStream` with a massive grain of salt. I just noticed, that ReaderStream has an option to increase the buffer size and I just used it differently/indirectly. So other than giving you more control, this **may not better than ReaderStream at all**. Maybe even worse since the guys that made `tokio_util` are surely way more knowledgeable about the area than I am.

EDIT 2: Just updated to Actix 4.4 and all seems good there as well. Doing 4 concurrent downloads and the frontend can still be loaded and websockets, etc work (even if slow at that point). Also first time I didn't need to update any of my actix code on updating. Really nice!
